library/ssl_msg.c

```diff
@@ -1425,20 +1425,43 @@ int mbedtls_ssl_decrypt_buf( mbedtls_ssl_context const *ssl,
          * Decrypt and authenticate
          */
 #if defined(MBEDTLS_USE_PSA_CRYPTO)
+        /*
+         * Workaround for lack of ChaCha/Poly in-place decryption in nrf_cc3xx
+         * by creating a stack variable to hold data to decrypt.
+         * Please see NCSDK-20188 for details on removing this workaround.
+         */
+        const size_t max_data_size = 128;
+        size_t plaintext_size = rec->buf_len - (data - rec->buf);
+        uint8_t out_data[max_data_size] = {0}:
+        if (transform->psa_alg == PSA_ALG_CHACHA20_POLY1305 &&
+            plaintext_size > max_data_size)
+        {
+            ret = MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL;
+            MBEDTLS_SSL_DEBUG_RET( 1, "psa_aead_decrypt (ChaCha/Poly)", ret );
+            return( ret );
+        }
+
         status = psa_aead_decrypt( transform->psa_key_dec,
                                transform->psa_alg,
                                iv, transform->ivlen,
                                add_data, add_data_len,
                                data, rec->data_len + transform->taglen,
-                               data, rec->buf_len - (data - rec->buf),
+                               out_data, plaintext_size,
                                &olen );
 
         if( status != PSA_SUCCESS )
         {
+            mbedtls_platform_zeroize(out_data, max_data_size);
             ret = psa_ssl_status_to_mbedtls( status );
             MBEDTLS_SSL_DEBUG_RET( 1, "psa_aead_decrypt", ret );
             return( ret );
         }
+
+        /* Copy data from out_data to data in case ChaCha/Poly was used. */
+        if (transform->psa_alg == PSA_ALG_CHACHA20_POLY1305)
+        {
+            memcpy(data, out_data, olen);
```

Comment:
This does not look right. out_data is always the argument given to psa_aead_decrypt, so the memcpy should always be done.

Or we should check for transform->psa_alg == CHACHA and do a different call to psa_aead_decrypt

Reply:
You are correct. Will fix

Reply:
Split into multi-part updates so temporary buffer is only used for ChaCha/Poly